### DISCUSSION TITLE
GAE with multiple graphs

### DISCUSSION THREAD

**Discussion by jangkj09** [pyg-team/pytorch_geometric#5649] [Q&A]
Hi, 

I'm trying to understand how to train a GAE (or VGAE) with multiple graphs. I see bits and pieces in the discussions and issues (https://github.com/pyg-team/pytorch_geometric/issues/458, https://github.com/pyg-team/pytorch_geometric/discussions/3769, https://github.com/pyg-team/pytorch_geometric/discussions/3046, https://github.com/pyg-team/pytorch_geometric/issues/1176, etc.) but nothing conclusive. Is there a tutorial or implementation that utilizes multiple graphs to train a GAE?

If not, do I need to simply do a RandomLinkSpit on each graph in a batch during the training loop as done in here (https://github.com/pyg-team/pytorch_geometric/issues/458)?

I'm also not sure  "integrate batch support in the decoder to fix (using the to_dense_batch method)" (https://github.com/pyg-team/pytorch_geometric/issues/458#issuecomment-506184328) means.

Could some guidance be provided? Unless there is already an implementation.

**Comment by rusty1s** [comment]
We sadly don't have an example for this yet, I am sorry. The process to do this would be

1. You call `RandomLinkSplit` on each `data` object before training.
2. You call `batched_negative_sampling` within your training loop to generate negative examples.

Everything else stays mostly the same as in our `link_pred.py` example. Happy to help you further. We can also work on this together in a small PR. Let me know if you are interested.

**Reply by jangkj09** [reply]
Sure, I would like to figure this out. It would be an opportunity to understand pyg more in depth as well. 
Just one quick question, would I call `RandomLinkSplit` on each `data` object before I create the dataloader? Or after?

I tried it after and during a training loop that examines the batches, but it seemed to have some weird behavior where some of the `data` examples lose their edges. I'm not sure if its my dataset or if its something else. I will look into this myself.

Either way, please let me know how to proceed with the PR.

Thanks!

**Reply by rusty1s** [reply]
The split needs to be fixed in advance, so what you can do is:
```python
dataset = MyDataset(..., transform=RandomLinkSplit(...))
train_dataset, val_dataset, test_dataset = zip(*[data for data in dataset])
```
